src/serde/object_cache.rs

```diff
@@ -0,0 +1,268 @@
+            None => None,
+            Some(left_value) => cache
+                .get_from_cache(&right)
+                .map(|right_value| 1 + left_value + right_value),
+        },
+        SExp::Atom(atom_buf) => {
+            let buf = allocator.buf(&atom_buf);
+            let lb = buf.len();
+            Some(if lb == 0 || (lb == 1 && buf[0] < 128) {
+                1
+            } else if lb < 0x40 {
+                1 + lb
+            } else if lb < 0x2000 {
+                2 + lb
+            } else if lb < 0x100000 {
+                3 + lb
+            } else if lb < 0x8000000 {
+                4 + lb
+            } else {
+                5 + lb
+            })
+        }
+    }
+}
+
+#[cfg(test)]
+use std::cmp::max;
+
+#[cfg(test)]
+use std::fmt::Debug;
+
+#[cfg(test)]
+use std::io::Cursor;
+
+#[cfg(test)]
+use hex::FromHex;
+
+#[cfg(test)]
+use crate::serde::de::node_from_stream;
+
+/// calculate the depth of a node. Used for tests
+
+#[cfg(test)]
+fn calculate_depth_simple(
+    cache: &mut ObjectCache<usize>,
+    allocator: &Allocator,
+    node: NodePtr,
+) -> Option<usize> {
+    match allocator.sexp(node) {
+        SExp::Pair(left, right) => match cache.get_from_cache(&left) {
+            None => None,
+            Some(left_value) => cache
+                .get_from_cache(&right)
+                .map(|right_value| 1 + max(*left_value, *right_value)),
+        },
+        SExp::Atom(_atom_buf) => Some(0),
+    }
+}
+
+#[cfg(test)]
+fn check_cached_function<T>(obj_as_hex: &str, expected_value: T, f: CachedFunction<T>)
+where
+    T: Clone + Eq + Debug,
+{
+    let mut allocator = Allocator::new();
+    let blob: Vec<u8> = Vec::from_hex(obj_as_hex).unwrap().into();
+    let mut cursor: Cursor<&[u8]> = Cursor::new(&blob);
+    let obj = node_from_stream(&mut allocator, &mut cursor).unwrap();
+    let mut oc = ObjectCache::new(&allocator, f);
+
+    assert_eq!(oc.get_from_cache(&obj), None);
+
+    oc.calculate(&obj);
+
+    assert_eq!(oc.get_from_cache(&obj), Some(&expected_value));
+
+    assert_eq!(oc.get_or_calculate(&obj).unwrap().clone(), expected_value);
+
+    assert_eq!(oc.get_from_cache(&obj), Some(&expected_value));
+
+    // do it again, but the simple way
+    let mut oc = ObjectCache::new(&allocator, f);
+    assert_eq!(oc.get_or_calculate(&obj).unwrap().clone(), expected_value);
+}
+
+#[test]
+fn test_depths_cache() {
+    let check = |a, b| check_cached_function(a, b, calculate_depth_simple);
+    check("01", 0); // 1
+    check("ff83666f6f83626172", 1); // (foo . bar)
+    check("ff83666f6fff8362617280", 2); // (foo bar)
+    check("ffff0102ff0304", 2); // ((1 . 2) . (3 . 4))
+    check("ff01ff02ff03ff04ff05ff0680", 6); // (1 2 3 4 5 6)
+}
+
+#[test]
+fn test_treehash() {
+    let check = |a, b| check_cached_function(a, Bytes32::from_hex(&b).unwrap(), treehash);
+    check(
+        "ff83666f6f83626172",
+        "c518e45ae6a7b4146017b7a1d81639051b132f1f5572ce3088a3898a9ed1280b",
```

Comment:
I think I wrote this before I decided I needed a `deserialize_wbr` function. It would be a lot easier to simply do a round-trip and make sure it's idempotent. I'll make that change.

Note that you can use `opc -H (foo . bar)` to quickly get a tree hash (and in fact, that's what I did).

Reply:
Done in @f70809ee3f33dce00327eb9a97139082bd1dc54d